### PR TITLE
Switch to lower contention ClockCache

### PR DESCRIPTION
**Pull request by benaadams**
## Changes

- Switch LruCache to ClockCache
    - Doesn't take any locks in the read path; only marking in bitmap as accessed using Interlocked
    - On Set it takes lock scanning the bitmap advancing a "clock" integer and setting each item as 0, until it comes to one already 0, which it evicts.

#### What types of changes does your code introduce?

- [x] Optimization

#### Requires testing

- [x] Yes

#### If yes, did you write tests?

- [x] Yes

### CODE REVIEW

File: src/Nethermind/Nethermind.Core/Caching/ClockCache.cs [hunk 1]
```diff
@@ -0,0 +1,248 @@
...
+        /// This doesn't hold true for ((n - 1) / 64) + 1, which equals 1.
+        ///
+        /// Usage:
+        /// GetInt32ArrayLengthFromBitLength(77): returns how many ints must be
+        /// allocated to store 77 bits.
+        /// </summary>
+        /// <param name="n"></param>
+        /// <returns>how many ints are required to store n bytes</returns>
+        private static int GetInt64ArrayLengthFromBitLength(int n) =>
+            (n - 1 + (1 << BitShiftPerInt64)) >>> BitShiftPerInt64;
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Core/Caching/ClockCache.cs:240]: Make this some helper method available for all caches?

File: src/Nethermind/Nethermind.Core/Caching/ClockCache.cs [hunk 2]
```diff
@@ -0,0 +1,248 @@
...
+        where TKey : struct, IEquatable<TKey>
+    {
+        private const int BitShiftPerInt64 = 6;
+        private readonly int _maxCapacity;
+        private readonly ConcurrentDictionary<TKey, LruCacheItem> _cacheMap;
+        private readonly McsLock _lock = new();
+        private readonly string _name;
+        private readonly TKey[] _keys;
+        private readonly long[] _accessedBitmap;
+        private readonly Queue<int> _freeOffsets = new();
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Core/Caching/ClockCache.cs:26]: This 3 collections and all the logic around them should be extracted to separate component

File: src/Nethermind/Nethermind.Core/Caching/ClockCache.cs [hunk 3]
```diff
@@ -0,0 +1,248 @@
...
+namespace Nethermind.Core.Caching
+{
+    public sealed class ClockCache<TKey, TValue>
+        where TKey : struct, IEquatable<TKey>
+    {
+        private const int BitShiftPerInt64 = 6;
+        private readonly int _maxCapacity;
+        private readonly ConcurrentDictionary<TKey, LruCacheItem> _cacheMap;
+        private readonly McsLock _lock = new();
+        private readonly string _name;
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Core/Caching/ClockCache.cs:23]: name is never used, lets remove it

**Review by LukaszRozmej — COMMENTED** [review]
Overall well done, but can we remove code duplication, by extracting a shared component that would do the access tracking?

File: src/Nethermind/Nethermind.Core/Caching/ClockCache.cs (hunk 3 above)
Comment (benaadams) [reply src/Nethermind/Nethermind.Core/Caching/ClockCache.cs]: Removed

File: src/Nethermind/Nethermind.Core/Caching/ClockCache.cs (hunk 2 above)
Comment (benaadams) [reply src/Nethermind/Nethermind.Core/Caching/ClockCache.cs]: Moved to base class

File: src/Nethermind/Nethermind.Core/Caching/ClockCache.cs (hunk 1 above)
Comment (benaadams) [reply src/Nethermind/Nethermind.Core/Caching/ClockCache.cs]: Moved to base class

File: src/Nethermind/Nethermind.Core/Caching/ClockKeyCacheNonConcurrent.cs [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+        long flags = 1L << position;
+
+        ref long accessedBitmapWord = ref Unsafe.Add(ref MemoryMarshal.GetArrayDataReference(HasBeenAccessedBitmap), offset);
+        bool accessed = (accessedBitmapWord & flags) != 0;
+        accessedBitmapWord &= ~flags;
+
+        return accessed;
+    }
+
+    private void MarkAccessedNonConcurrent(int position)
```
Comment (Scooletz) [inline src/Nethermind/Nethermind.Core/Caching/ClockKeyCacheNonConcurrent.cs:128]: I'd push it to the the base, leave it to the caller to use.

**Review by Scooletz — CHANGES_REQUESTED** [review]
The Clock bit operations is pretty similar to the `BitFilter` in `Paprika` 😅 
The clock is simple and easy to reason about.

I believe we can make atomics simpler. Provided comments.

File: src/Nethermind/Nethermind.Core/Caching/ClockCacheBase.cs [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+        HasBeenAccessedBitmap.AsSpan().Clear();
+    }
+
+    protected bool ClearAccessed(int position)
+    {
+        uint offset = (uint)position >> BitShiftPerInt64;
+        long flags = 1L << position;
+
+        ref long accessedBitmapWord = ref Unsafe.Add(ref MemoryMarshal.GetArrayDataReference(HasBeenAccessedBitmap), offset);
+        bool accessed = (accessedBitmapWord & flags) != 0;
```
[the remark this answers is not in the document]
Comment (benaadams) [reply src/Nethermind/Nethermind.Core/Caching/ClockCacheBase.cs]: Changed

File: src/Nethermind/Nethermind.Core/Caching/ClockCacheBase.cs [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+
+        if (accessed)
```
[the remark this answers is not in the document]
Comment (benaadams) [reply src/Nethermind/Nethermind.Core/Caching/ClockCacheBase.cs:49]: Done

File: src/Nethermind/Nethermind.Core/Caching/ClockKeyCacheNonConcurrent.cs [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+    }
+
+    public bool Contains(TKey key)
+    {
+        return _cacheMap.ContainsKey(key);
+    }
+
+    public int Count => _cacheMap.Count;
+
+    private bool ClearAccessedNonConcurrent(int position)
```
[the remark this answers is not in the document]
Comment (benaadams) [reply src/Nethermind/Nethermind.Core/Caching/ClockKeyCacheNonConcurrent.cs]: Done

File: src/Nethermind/Nethermind.Core/Caching/ClockKeyCacheNonConcurrent.cs (hunk 4 above)
Comment (benaadams) [reply src/Nethermind/Nethermind.Core/Caching/ClockKeyCacheNonConcurrent.cs]: Done

File: src/Nethermind/Nethermind.Core/Caching/ClockCacheBase.cs [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+    {
+        uint offset = (uint)position >> BitShiftPerInt64;
+        long flags = 1L << position;
+
+        ref long accessedBitmapWord = ref Unsafe.Add(ref MemoryMarshal.GetArrayDataReference(HasBeenAccessedBitmap), offset);
+
+        long current = Volatile.Read(ref accessedBitmapWord);
+        while (true)
+        {
+            long previous = Interlocked.CompareExchange(ref accessedBitmapWord, current | flags, current);
```
[the remark this answers is not in the document]
Comment (benaadams) [reply src/Nethermind/Nethermind.Core/Caching/ClockCacheBase.cs]: Done

**Review by Scooletz — APPROVED** [review]
❤️

**Review by LukaszRozmej — APPROVED** [review]

**Pull request merged**